### PR TITLE
Add profile event `SelectQueriesWithPrimaryKeyUsage`

### PR DESCRIPTION
**Pull request by 0xfei**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added profile event `SelectQueriesWithPrimaryKeyUsage` which indicates how many SELECT queries use the primary key to evaluate the WHERE clause.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 2daaf5362fd6274c2f41c517c0041f692eae6830 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64492/2daaf5362fd6274c2f41c517c0041f692eae6830/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by 0xfei** [comment]
02968_file_log_multiple_read	FAIL	12.3
2024-05-28 12:52:02 --- /ClickHouse/tests/queries/0_stateless/02968_file_log_multiple_read.reference	2024-05-28 12:36:38.814902628 +0200
2024-05-28 12:52:02 +++ /ClickHouse/tests/queries/0_stateless/02968_file_log_multiple_read.stdout	2024-05-28 12:52:02.523438528 +0200
2024-05-28 12:52:02 @@ -18,13 +18,3 @@
2024-05-28 12:52:02  8
2024-05-28 12:52:02  9
2024-05-28 12:52:02  10
2024-05-28 12:52:02 -11
2024-05-28 12:52:02 -12
2024-05-28 12:52:02 -13
2024-05-28 12:52:02 -14
2024-05-28 12:52:02 -15
2024-05-28 12:52:02 -16
2024-05-28 12:52:02 -17
2024-05-28 12:52:02 -18
2024-05-28 12:52:02 -19
2024-05-28 12:52:02 -20
2024-05-28 12:52:02 
2024-05-28 12:52:02 
2024-05-28 12:52:02 
2024-05-28 12:52:02 Database: test_mrevuoj2

@nikitamikhaylov Could you please help confirm if this failed test is related to this commit?

File: docs/en/operations/system-tables/query_log.md [hunk 1]
```diff
@@ -151,6 +152,7 @@ query_kind:
 databases:                             []
 tables:                                []
 columns:                               []
+columns_in_where:                      ['type']
```
Comment (rschu1ze) [inline docs/en/operations/system-tables/query_log.md:155]: Seems weird that `type` appears in field `columns_in_where` but not in `columns`.

File: docs/en/operations/system-tables/query_log.md [hunk 2]
```diff
@@ -61,6 +61,7 @@ Columns:
 - `databases` ([Array](../../sql-reference/data-types/array.md)([LowCardinality(String)](../../sql-reference/data-types/lowcardinality.md))) — Names of the databases present in the query.
 - `tables` ([Array](../../sql-reference/data-types/array.md)([LowCardinality(String)](../../sql-reference/data-types/lowcardinality.md))) — Names of the tables present in the query.
 - `columns` ([Array](../../sql-reference/data-types/array.md)([LowCardinality(String)](../../sql-reference/data-types/lowcardinality.md))) — Names of the columns present in the query.
+- `columns_in_where` ([Array](../../sql-reference/data-types/array.md)([LowCardinality(String)](../../sql-reference/data-types/lowcardinality.md))) — Names of the columns present in the query where conditions.
```
Comment (rschu1ze) [inline docs/en/operations/system-tables/query_log.md:64]: To be honest, I don't like the new field. It seems not systematic. Someone else may add `columns_in_select`, `columns_in_order_by`, `columns_in_group_by` etc., and then do the same for `databases`, `tables`, `projections` etc. In the end, the entire query_log will blow up.

Comment (0xfei) [reply docs/en/operations/system-tables/query_log.md:64]: In many scenarios involving complex tables and BI-generated complex queries, users are often unsure about setting appropriate primary keys and ORDER BY keys. We aim to automatically adjust the ORDER BY clause based on user queries to optimize query efficiency. Therefore, columns_in_where are extremely useful for us. Alternatively, could you provide suggestions on how to implement this? @rschu1ze

Comment (rschu1ze) [reply docs/en/operations/system-tables/query_log.md:64]: The use case you describe is about recommending a primary key for a table based on the workload (SELECT queries).

This is a well-known problem and the best solution is to observe (e.g. keep statistics about) filter operations during query execution. The problem with using the raw query strings to suggest a primary key is that the query optimizer may (in principle) rewrite/optimize the query in arbitrary ways such that the filter columns in the SELECT query (WHERE clause) no longer corresponds to the actually performed filter operations. Unfortunately, ClickHouse does not have such a feature right now, it would be cool to implement that of course.

Comment (0xfei) [reply docs/en/operations/system-tables/query_log.md:64]: YES, We're currently working on it and hope to add 'columns_in_where' to avoid a large amount of redundant and incorrect offline query analysis.

Comment (0xfei) [reply docs/en/operations/system-tables/query_log.md:64]: unfortunately, this CI has been a bit problematic, and I'm still working on it. Could you please continue with the review...

Comment (rschu1ze) [reply docs/en/operations/system-tables/query_log.md:64]: I am personally not a big fan of extending `query_log`, it will be a quick and dirty solution for a problem that needs to be solved more comprehensively. But that is just my opinion. Other reviewers may have a different view.

File: src/Common/ProfileEvents.cpp [hunk 3]
```diff
@@ -190,8 +190,13 @@
     \
     M(ReplicaPartialShutdown, "How many times Replicated table has to deinitialize its state due to session expiration in ZooKeeper. The state is reinitialized every time when ZooKeeper is available again.") \
     \
+    M(PrimaryKeyNotUsed, "Check if primary key is used in this query.") \
```
Comment (rschu1ze) [inline src/Common/ProfileEvents.cpp:193]: The new ProfileEvents are orthogonal to the extension of `sytem.query_log`. Could we split either of these into a separate PR?

Comment (0xfei) [reply src/Common/ProfileEvents.cpp:193]: ok

File: docs/en/operations/system-tables/query_log.md (hunk 2 above)
Comment (0xfei) [reply docs/en/operations/system-tables/query_log.md:64]: Yes, I'm thinking of another way, planning to submit a PR to transform 'columns' into a MAP, just like ProfileEvents map, allowing clickhouse to record more information without affecting current usage or bloating the query_log.

**Comment by 0xfei** [comment]
Could you please review the code? @nikitamikhaylov

**Comment by 0xfei** [comment]
Could you please review the code? @rschu1ze @nikitamikhaylov

**Comment by rschu1ze** [comment]
Sorry, haven't had time yet. Will take a look today.

File: src/Common/ProfileEvents.cpp [hunk 4]
```diff
@@ -190,8 +190,13 @@
...
+    M(TotalParts, "Total number of data parts in a MergeTree table.") \
+    M(PartitionParts, "Number of data parts selected by partition in a MergeTree table.") \
     M(SelectedParts, "Number of data parts selected to read from a MergeTree table.") \
     M(SelectedRanges, "Number of (non-adjacent) ranges in all data parts selected to read from a MergeTree table.") \
+    M(TotalMarks, "Total number of marks (index granules) in a MergeTree table.") \
```
Comment (rschu1ze) [inline src/Common/ProfileEvents.cpp:198]: About TotalMarks and PrimaryKeyMarks: We have a similar problem here as above. Updating statistics while running queries is by definition not able to compute the total number of marks in a MergeTree Table.

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp [hunk 5]
```diff
@@ -2002,8 +2011,12 @@ void ReadFromMergeTree::initializePipeline(QueryPipelineBuilder & pipeline, cons
                 Context::QualifiedProjectionName{.storage_id = data.getStorageID(), .projection_name = storage_snapshot->projection->name});
     }
 
+    ProfileEvents::increment(ProfileEvents::TotalParts, result.total_parts);
```
Comment (rschu1ze) [inline src/Processors/QueryPlan/ReadFromMergeTree.cpp:2014]: I am really confused now: Four new metrics are introduced, and by the look of things they all take on the same value. Why would one introduce them in the first place? Can't users simply use `SelectedParts`, for example?

Comment (rschu1ze) [reply src/Processors/QueryPlan/ReadFromMergeTree.cpp:2014]: EDIT: I see ... the values are different, sorry.

File: src/Common/ProfileEvents.cpp [hunk 6]
```diff
@@ -190,8 +190,13 @@
     \
     M(ReplicaPartialShutdown, "How many times Replicated table has to deinitialize its state due to session expiration in ZooKeeper. The state is reinitialized every time when ZooKeeper is available again.") \
     \
+    M(PrimaryKeyNotUsed, "Check if primary key is used in this query.") \
+    M(TotalParts, "Total number of data parts in a MergeTree table.") \
+    M(PartitionParts, "Number of data parts selected by partition in a MergeTree table.") \
```
[the remark this answers is not in the document]
Comment (0xfei) [reply src/Common/ProfileEvents.cpp:195]: The number of data parts and the number of partitions should be different. Each partition usually has multiple parts, and I think users can understand this concept.

File: src/Common/ProfileEvents.cpp [hunk 7]
```diff
@@ -190,8 +190,13 @@
     \
     M(ReplicaPartialShutdown, "How many times Replicated table has to deinitialize its state due to session expiration in ZooKeeper. The state is reinitialized every time when ZooKeeper is available again.") \
     \
+    M(PrimaryKeyNotUsed, "Check if primary key is used in this query.") \
+    M(TotalParts, "Total number of data parts in a MergeTree table.") \
```
[the remark this answers is not in the document]
Comment (0xfei) [reply src/Common/ProfileEvents.cpp:194]: I will try to improve the description.

File: src/Common/ProfileEvents.cpp (hunk 4 above)
Comment (0xfei) [reply src/Common/ProfileEvents.cpp:198]: seems I have some misconceptions about profile events...

File: src/Common/ProfileEvents.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xfei) [reply src/Common/ProfileEvents.cpp:193]: ok...

**Comment by 0xfei** [comment]
Hi, I have reviewed the definition and usage of profile events in ClickHouse,  the previous code was indeed poorly written... please help me review it again, thx @rschu1ze

File: src/Interpreters/InterpreterSelectQueryAnalyzer.cpp [hunk 8]
```diff
@@ -226,6 +233,11 @@ BlockIO InterpreterSelectQueryAnalyzer::execute()
     if (!select_query_options.ignore_quota && select_query_options.to_stage == QueryProcessingStage::Complete)
         result.pipeline.setQuota(context->getQuota());
 
+
+    /// Increase the SelectQueriesWithPrimaryKey profile.
```
Comment (rschu1ze) [inline src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:237]: I see three problems with metric `SelectQueriesWithPrimaryKey` that will be hard to solve:

1. File InterpreterSelectQueryAnalyzer.cpp is the most central place of the Analyzer component of ClickHouse. Before this PR it was not cluttered with any ProfileEvent updates, now it is. ProfileEvent updates for similar counters (the ones in l. 10-17 of ProfileEvents.cpp) are implemented in Planner.cpp. Why is that not the case for SelectQueriesWithPrimaryKey?
2. Looking at the names and meaning (according to the comments) of the profile events in l. 10 - 17, they are all about counting queries of a specific type like SELECT and INSERT. Counting the number of queries that use the primary key breaks the scheme and introduces an inconsistency. The counter should be renamed to "SelectQueriesWithPrimaryKeyUsage" and moved before "AsyncInsertQuery". That is also not great but at least a little improvement.
3. The usage of the Context to store the metric result is unusual, error-prone, and further clutters the already too messy Context. We should keep it simple and not store ProfileEvents in the context. As far as I see, the context is required to get a per-query usage statistics (i.e. across multiple accessed parts) for the query. Could we not simply increase the metric by 1 in l. 1642 of ReadFromMergeTree.cpp and not update the context? The value will then be > 1 for a query but I think that is still okay?

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp [hunk 9]
```diff
@@ -117,8 +117,10 @@ bool restorePrewhereInputs(PrewhereInfo & info, const NameSet & inputs)
 
 namespace ProfileEvents
 {
+    extern const Event SelectedPartsByPartitionKey;
```
Comment (rschu1ze) [inline src/Processors/QueryPlan/ReadFromMergeTree.cpp:120]: Sorting:

```cpp
extern const Event SelectedMarks;
extern const Event SelectedMarksByPrimaryKey;
extern const Event SelectedParts;
extern const Event SelectedPartsByPartitionKey;
extern const Event SelectedRanges;
```

File: tests/queries/0_stateless/03164_parts_filter_profiles_test.sql [hunk 10]
```diff
@@ -0,0 +1,25 @@
+set log_queries=1;
+set log_queries_min_type='QUERY_FINISH';
+
+DROP TABLE IF EXISTS 03164_query_log_test_parts_filter_profiles;
+CREATE TABLE 03164_query_log_test_parts_filter_profiles
+(
+    a Int64, b Int64, c Int64,
+    INDEX idx_c(c) TYPE minmax
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03164_parts_filter_profiles_test.sql:8]: Is it right that the `minmax` index does not affect the value of the profile events? If yes, then let's remove it because it distracts from the actual purpose of the test.

File: tests/queries/0_stateless/03164_parts_filter_profiles_test.sql [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+) ENGINE = MergeTree PARTITION BY a ORDER BY b SETTINGS index_granularity=10;
+
+INSERT INTO 03164_query_log_test_parts_filter_profiles SELECT number, number+1000, number+10000 FROM numbers(100);
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03164_parts_filter_profiles_test.sql:11]: Because of `PARTITION BY a` (l. 9) we'll obtain 100 partitions with a single row each, right? That is not realistic at all.

Let's remove the partition spec from l. 9, and instead create two parts (in a single implicit partition):

```sql
INSERT INTO 03164_query_log_test_parts_filter_profiles SELECT number, number+1000, number+10000 FROM numbers(100);
INSERT INTO 03164_query_log_test_parts_filter_profiles SELECT number, number+1000, number+10000 FROM numbers(100);
```

EDIT: Just a single test query is not comprehensive testing. Needs tests for:
- the basic case: no partitions, two parts
- a test with partitions: two partitions with two parts each
- a test with projections: are the metrics properly updated if a projection exists for a table?
- a test where no primary key is used

File: tests/queries/0_stateless/03164_parts_filter_profiles_test.sql [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+SELECT
+    ProfileEvents.Values[indexOf(ProfileEvents.Names, 'SelectedPartsByPartitionKey')] as t1,
+    ProfileEvents.Values[indexOf(ProfileEvents.Names, 'SelectedMarksByPrimaryKey')] as t2,
+    ProfileEvents.Values[indexOf(ProfileEvents.Names, 'SelectQueriesWithPrimaryKey')] as t3
+FROM
+    system.query_log
+WHERE
+    current_database=currentDatabase() and
+    query = 'select count(*) from 03164_query_log_test_parts_filter_profiles where a between 1 and 30 and b between 1010 and 1020 and c between 10003 and 10013;';
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03164_parts_filter_profiles_test.sql:25]: At the end of the test, let's drop the table we created above in l. 5.

File: tests/queries/0_stateless/03164_parts_filter_profiles_test.sql [hunk 13]
```diff
@@ -0,0 +1,25 @@
+set log_queries=1;
+set log_queries_min_type='QUERY_FINISH';
+
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03164_parts_filter_profiles_test.sql:3]: The test name is too generic, developers will have no clue about the purpose of the test by looking only at its file name and will have to read it to understand. It also lacks a basic comment what is being tested. Ideally, please rename to something less generic. Please add a comment here, e.g. 

```sql
-- Tests ProfileEvents "SelectedPartsByPartitionKey", "SelectedPartsByPartitionKey", "SelectedMarksByPrimaryKey"

File: src/Common/ProfileEvents.cpp [hunk 14]
```diff
@@ -190,8 +191,10 @@
     \
     M(ReplicaPartialShutdown, "How many times Replicated table has to deinitialize its state due to session expiration in ZooKeeper. The state is reinitialized every time when ZooKeeper is available again.") \
     \
+    M(SelectedPartsByPartitionKey, "Number of data parts selected by partition in a MergeTree table.") \
```
Comment (rschu1ze) [inline src/Common/ProfileEvents.cpp:194]: Users will wonder why the generally more useful and general metrics (SelectedParts) is listed after a more specialized metric (SelectedPartsByPrimaryKey). 

Same in l. 197/198.

I know that the description (l. 194) was updated, but I still don't understand (and users will not understand too), what the metric represents. Especially "ByPartitionKey" is confusing, does that mean "ByPrimaryKey"??

Comment (0xfei) [reply src/Common/ProfileEvents.cpp:194]: I didn't come up with a better way to describe SelectedPartsByPartitionKey, deleted it for now.

**Comment by 0xfei** [comment]
split it into multiple PRs. please take a look @rschu1ze

**Comment by rschu1ze** [comment]
Which other PRs did you push exactly?

**Comment by 0xfei** [comment]
> Which other PRs did you push exactly?

The others haven't been pushed up yet, This one is about the PrimaryKeyUsage.

**Review by rschu1ze — APPROVED** [review]

**Comment by 0xfei** [comment]
Is it okay to merge? @rschu1ze

**Comment by rschu1ze** [comment]
Integration Tests (aarch64) [5/6]: #65028
Stateless Tests (debug, S3 Storage) [3/6]: #65472
AST Fuzzer (debug): 
Stateless Tests (tsan) [3/5]: #65310
